Packet++/header/WireGuardLayer.h

```diff
@@ -0,0 +1,253 @@
+	 * Represents the Handshake Initiation message
+	 */
+#pragma pack(push, 1)
+	struct wg_handshake_initiation
+	{
+		wg_common_header common;               ///< Common header for all WireGuard messages
+		uint32_t senderIndex;                  ///< Sender index
+		uint8_t initiatorEphemeral[32];        ///< Initiator's ephemeral public key
+		uint8_t encryptedInitiatorStatic[48];  ///< Encrypted initiator's static key
+		uint8_t encryptedTimestamp[28];        ///< Encrypted timestamp
+		uint8_t mac1[16];                      ///< MAC1 field
+		uint8_t mac2[16];                      ///< MAC2 field
+	};
+#pragma pack(pop)
+
+	/**
+	 * @struct wg_handshake_response
+	 * Represents the Handshake Response message
+	 */
+#pragma pack(push, 1)
+	struct wg_handshake_response
+	{
+		wg_common_header common;         ///< Common header for all WireGuard messages
+		uint32_t senderIndex;            ///< Sender index
+		uint32_t receiverIndex;          ///< Receiver index
+		uint8_t responderEphemeral[32];  ///< Responder's ephemeral public key
+		uint8_t encryptedEmpty[16];      ///< Encrypted empty field
+		uint8_t mac1[16];                ///< MAC1 field
+		uint8_t mac2[16];                ///< MAC2 field
+	};
+#pragma pack(pop)
+
+	/**
+	 * @struct wg_cookie_reply
+	 * Represents the Cookie Reply message
+	 */
+#pragma pack(push, 1)
+	struct wg_cookie_reply
+	{
+		wg_common_header common;      ///< Common header for all WireGuard messages
+		uint32_t receiverIndex;       ///< Receiver index
+		uint8_t nonce[24];            ///< Nonce field
+		uint8_t encryptedCookie[32];  ///< Encrypted cookie
+	};
+#pragma pack(pop)
+
+	/**
+	 * @struct wg_transport_data
+	 * Represents the Transport Data message
+	 */
+#pragma pack(push, 1)
+	struct wg_transport_data
+	{
+		wg_common_header common;   ///< Common header for all WireGuard messages
+		uint32_t receiverIndex;    ///< Receiver index
+		uint64_t counter;          ///< Counter field
+		uint8_t encryptedData[0];  ///< Flexible array member for encrypted data
+	};
+#pragma pack(pop)
+
+	/**
+	 * @class WireGuardLayer
+	 * Represents a WireGuard protocol layer
+	 */
+	class WireGuardLayer : public Layer
+	{
+	public:
+		/**
+		 * Constructs a WireGuardLayer object.
+		 *
+		 * @param data Pointer to the raw data representing the WireGuard layer
+		 * @param dataLen Length of the data
+		 * @param prevLayer Pointer to the previous layer in the packet (if any)
+		 * @param packet Pointer to the packet this layer belongs to
+		 */
+		WireGuardLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
+		    : Layer(data, dataLen, prevLayer, packet, Wireguard)
+		{}
+
+		/**
+		 * Gets a pointer to the Handshake Initiation message.
+		 *
+		 * @return Pointer to the Handshake Initiation message, or nullptr if data is invalid
+		 */
+		const wg_handshake_initiation* getHandshakeInitiation() const
+		{
+			return reinterpret_cast<const wg_handshake_initiation*>(m_Data);
+		}
```

Comment:
A single packet can include one of the 4 possible messages or another unknown message (which may indicate the packet is invalid). But what if the user doesn't know which type of packet it is? How would he know which getter to call?

We usually solve this problem in one of two ways:
1. Create a base and derived classes for each message type. A good example is [`BgpLayer`](https://github.com/seladb/PcapPlusPlus/blob/master/Packet%2B%2B/header/BgpLayer.h)
2. Include all messages in the base class, closer to this implementation like we do in [`IcmpLayer`](https://github.com/seladb/PcapPlusPlus/blob/master/Packet%2B%2B/header/IcmpLayer.h), although this is less common and we rarely do it lately

If you choose option (2), which is closer to your implementation (but less OOP IMO), I'd expect the following:
- Before doing `reinterpret_cast` check if the message is of the right type and if the length is valid, otherwise return `nullptr`
- Provide another method to get the type, like: `WireGuardMessageType getWireGuardType()` so users can know the packet type before calling the relevant method

Reply:
Thank you for the feedback. I have implemented the first approach by creating a base class and derived classes for each message type, similar to the BgpLayer implementation.